README.md
# icmp_watch
Send batch requests for ICMP and show results in a console window to monitor availability of hosts.

![console](console.png "console")


## Limitations

 * Uses ICMP sockets, which means that sysctl net.ipv4.ping_group_range needs to be set properly. (My Ubuntu 22.04 was configured correctly for this, by default.)

## Usage

`$ ./icmp_watch [option]... hostname...`

(Multiple hostnames can be specified.)

Valid options are:

* `-i` or `--interval`: specify how long in seconds to wait for replies (real numbers, e.g. 1.5 are allowed, default is 1 second)
* `-4` and `-6`: restrict to only using IPv4 or IPv6 respectively
* `-h` or `--help`: show help text

Press Q or ESC to stop monitoring.


## Function

This tool will batch test availability by sending out N ICMP-requests, and then wait up to 1 second for N replies.

Hosts that do not reply will be marked in RED, others in GREEN with the response time shown in milliseconds.

Efficiently implemented in C, zero polling, two sockets for all N hosts (one for all IPv4 connections, one for all IPv6 connections).


## Features

 * Super lightweight: uses 0.0% CPU/MEM on my machine.
 * Single file C source.
 * MIT license.


## Troubleshoot

If the tool fails on permissions when creating the socket, please do:

`$ sysctl -w net.ipv4.ping_group_range="0 2147483647"` to allow users to create ICMP sockets. You can use "0 0" range if you want this for root only.


## Author

The icmp_watch tool is made by Abraham Stolk.



debian/changelog
icmpwatch (1.0-1) kinetic; urgency=medium

  * Initial release.

 -- Bram Stolk <bram.stolk@canonical.com>  Wed, 07 Sep 2022 15:06:00 -0700


icmpwatch.c
// icmp_watch.c
//
// By Abraham Stolk.
// MIT LICENSE

#include <stdio.h>
#include <errno.h>
#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <termios.h>	// for tcsetattr()
#include <netdb.h>	// for getattrinfo()
#include <time.h>	// for nanosleep()
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip_icmp.h>
#include <arpa/inet.h>
#include <sys/select.h>
#include <getopt.h>		// for getopt_long()

// NOTE:
// To allow root to use icmp sockets, run:
// $ sysctl -w net.ipv4.ping_group_range="0 0"
// To allow all users to use icmp sockets, run:
// $ sysctl -w net.ipv4.ping_group_range="0 2147483647"

#define ESC	    "\x1B"

#define RESETALL    ESC "[0m"

#define CURSORHOME  ESC "[H"

#define CLEARSCREEN ESC "[H" ESC "[2J" ESC "[3J"

#define FGWHT	    ESC "[1;37m"

#define BGRED	    ESC "[1;41m"

#define BGGRN	    ESC "[1;42m"

static struct termios orig_termios;    // The terminal settings before we modified it.

void disableRawMode()
{
	tcsetattr(STDIN_FILENO, TCSAFLUSH, &orig_termios);
}

void enableRawMode()
{
	tcgetattr(STDIN_FILENO, &orig_termios);
	atexit(disableRawMode);
	struct termios raw = orig_termios;
	raw.c_lflag &= ~(ECHO);	     // Don't echo key presses.
	raw.c_lflag &= ~(ICANON);    // Read by char, not by line.
	raw.c_cc[VMIN] = 0;	     // No minimum nr of chars.
	raw.c_cc[VTIME] = 0;	     // No waiting time.
	tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw);
}

static int ping_all(int cnt, struct in_addr* destinations, int* response_times, int* errors, struct timeval* timeout)
{
	static int sequence = 0;
	const int seq = sequence++;        // the sequence number we will use for this run.
	const int waittime = (int) (timeout->tv_sec * 1000000 + timeout->tv_usec);
	int sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP);
	if (sock < 0)
	{
		perror("socket");
		fprintf(stderr,
			"To allow root to use icmp sockets, run:\n"
			"$ sudo sysctl -w net.ipv4.ping_group_range=\"0 0\"\n"
			"To allow all users to use icmp sockets, run:\n"
			"$ sudo sysctl -w net.ipv4.ping_group_range=\"0 2147483647\"\n"
		);
		exit(4);
	}

	for (int i = 0; i < cnt; ++i) {
		response_times[i] = -1;	  	// what we return if we did not get a reply.
		errors[i] = 0;			  	// what we return if there wasn’t an error
	}

	// Send an ICMP request to each destination.
	struct icmphdr icmp_hdr;
	unsigned char txdata[256];
	const int payloadsz = 10;
	memcpy(txdata + sizeof icmp_hdr, "icmp_watch", payloadsz);    // icmp payload
	for (int i = 0; i < cnt; ++i)
	{
		struct sockaddr_in addr;
		memset(&addr, 0, sizeof addr);
		addr.sin_family = AF_INET;
		addr.sin_addr = destinations[i];

		memset(&icmp_hdr, 0, sizeof icmp_hdr);
		icmp_hdr.type = ICMP_ECHO;
		icmp_hdr.un.echo.id = 0xbeef;

		icmp_hdr.un.echo.sequence = seq;
		memcpy(txdata, &icmp_hdr, sizeof icmp_hdr);
		int rc = sendto(sock, txdata, sizeof icmp_hdr + payloadsz, 0, (struct sockaddr*) &addr, sizeof addr);
		if (rc <= 0)
		{
			errors[i] = errno;
		}
	}

	fd_set read_set;
	memset(&read_set, 0, sizeof read_set);
	FD_SET(sock, &read_set);

	int num_replies = 0;
	while (num_replies < cnt)
	{
		// wait for a reply with a timeout
		const int rc0 = select(sock + 1, &read_set, NULL, NULL, timeout);
		if (rc0 == 0)
		{
			// Timed out without a reply.
			close(sock);
			sock = 0;
			return num_replies;
		}
		else if (rc0 < 0)
		{
			close(sock);
			sock = 0;
			perror("select");
			return -1;
		}

		unsigned char rcdata[256];
		struct icmphdr rcv_hdr;
		struct sockaddr_in other_addr;
		socklen_t other_addr_len = sizeof(other_addr);
		const int rc1 = recvfrom(sock, rcdata, sizeof rcdata, 0, (struct sockaddr*) &other_addr, &other_addr_len);
		if (rc1 <= 0)
		{
			perror("recvfrom");
			exit(5);
		}
		if (rc1 < (int) sizeof(rcv_hdr))
			exit(6);			// ICMP packet was too short.
		memcpy(&rcv_hdr, rcdata, sizeof rcv_hdr);
		assert(rcv_hdr.type == ICMP_ECHOREPLY);
		if (rcv_hdr.un.echo.sequence == seq)	// The sequence number should match, otherwise it's not a valid response.
		{
			const struct in_addr send_addr = other_addr.sin_addr;
			int idx = -1;
			// Look up which host sent us this reply.
			for (int j = 0; j < cnt; ++j)
				if (destinations[j].s_addr == send_addr.s_addr)
					idx = j;
			assert(idx >= 0);
			const int timeleft = (int) (timeout->tv_sec * 1000000 + timeout->tv_usec);
			response_times[idx] = waittime - timeleft;
			num_replies += 1;
		}
	}
	if (close(sock) < 0)
		perror("close(socket)");
	sock = 0;
	return num_replies;
}

// Resolves a set of hostnames to IPv4 numbers.
static int get_ip_addresses(int cnt, char** hosts, int args_left, struct in_addr* ips)
{
	struct addrinfo hints;
	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_INET;
	hints.ai_socktype = SOCK_DGRAM;
	for (int i = 0; i < cnt; ++i)
	{
		struct addrinfo* infos;
		const char* host = hosts[args_left + i];
		const int rv = getaddrinfo(host, 0, &hints, &infos);
		if (rv)
		{
			fprintf(stderr, "getaddrinfo() %s\n", gai_strerror(rv));
			return i;
		}
		for (struct addrinfo* inf = infos; inf != 0; inf = inf->ai_next)
		{
			struct sockaddr_in* addr = (struct sockaddr_in*) inf->ai_addr;
			ips[i] = addr->sin_addr;
		}
		freeaddrinfo(infos);
	}
	return cnt;
}

void print_help(char *progname) {
	// Restyled the help to more closely match the --help text for mv, etc.
	printf("Usage: %s [option]... destination_ip...\n"
		   "Send batch requests for ICMP and show the results\nPress q or escape to exit\n\n"
		   "  -i, --interval=INTERVAL\tspecify how long in seconds to wait for replies (real numbers, e.g. 1.5 are allowed)\n"
		   "  -h, --help\t\t\tshow this help\n", progname);
}

int main(int argc, char* argv[])
{
	struct timeval default_timeout = {1, 0};    // seconds, microseconds.
	
	// Parse command line options (we'll break out of the loop)
	while(1) {
		int option_index = 0;
		int c;
		static struct option long_options[] = {
			{"interval", required_argument, 0, 'i'},
			{"help", no_argument, 0, 'h'},
			{0, 0, 0, 0} 	// default option (for unknown options)
		};
		
		c = getopt_long(argc, argv, "i:h", long_options, &option_index);
		
		if (c == -1) {
			break; // no more options
		}
		
		//printf("Right now c is %i (%c)\n", c, c);
		switch(c) {
			case 0:
				printf("Unknown option %s\n", long_options[option_index].name);
				break;
			case 'i':
				// Convert the argument to --interval to a (double) float, then convert it to a timespec struct
				if(optarg) {
					double interval_double;
					errno = 0;		// Set errno to 0 so we can detect errors
					interval_double = strtod(optarg, NULL);
					if(errno != 0) {
						perror("Converting interval");
						exit(1);
					}
					// Set the specified interval as the default timeout
					default_timeout.tv_sec = (int) interval_double;
					default_timeout.tv_usec = (int) (interval_double * 1000000);
					default_timeout.tv_usec = default_timeout.tv_usec % 1000000;
					break;
				} else {
					fprintf(stderr, "-i/--interval needs an argument\n");
					exit(1);
				}
			case 'h':
				print_help(argv[0]);
				exit(0);
			default:
				//fprintf(stderr, "getopt_long returned character code 0x%x\n", c);
				;
		}
	} // Finished parsing command line options

	if (optind < argc) {
		//printf("%i non-option argv elements?\n", argc - optind);
	} else {
		print_help(argv[0]);
		return 1;
	}
	
	const int cnt = argc - optind;    // Every argument left after taking away the options is a hostname.
	struct in_addr dst[cnt];     // The IP numbers of the hosts.
	int response_times[cnt];     // The response time for each host we ping.
	int errors[cnt];             // The errno(3) for each host (0 if no error)

	fprintf(stderr, "Looking up %d ip numbers...", cnt);
	fflush(stderr);
	const int num = get_ip_addresses(cnt, argv, optind, dst);
	fprintf(stderr, "DONE\n");
	if (num != cnt)
	{
		fprintf(stderr, "Could not resolve all hostnames. Aborting.\n");
		exit(2);
	}

	enableRawMode();    // Don't echo keyboard characters, don't buffer them.

	int done = 0;
	int spaceForHostname = 19;	// Default spacing is 19 characters. (it was 20 before, but it'll get incremented later to ensure that there's always at least one space between the hostname and the result)
	
	// Check if any of the hostnames are longer than 20 characters and up spaceForHostname if the are
	for (int i = 0; i < cnt; i++) {
		/* this line will need changing if the cli options patch is merged */
		int lengthOfThisHostname = strlen(argv[1 + i]);
		if(lengthOfThisHostname > spaceForHostname) {
			spaceForHostname = lengthOfThisHostname;
		}
	}
	// Increment space for hostname by one so that there will always be at least one space between it and the result
	spaceForHostname++;
	
	while (!done)
	{
		// When ESC or Q is pressed, we should terminate.
		char c;
		const int numr = read(STDIN_FILENO, &c, 1);
		if (numr == 1 && (c == 27 || c == 'q' || c == 'Q'))
			done = 1;
		struct timeval timeout = default_timeout;
		ping_all(cnt, dst, response_times, errors, &timeout);
		fprintf(stdout, CLEARSCREEN);
		for (int i = 0; i < cnt; ++i)
		{
			const int t = response_times[i];
			const int e = errors[i];
			fprintf(stdout, "%-*s", spaceForHostname, argv[optind + i]);
			if (t < 0)
				if (e != 0)
					fprintf(stdout, FGWHT BGRED "   ERROR" RESETALL " (%s)\n", strerror(e));
				else
					fprintf(stdout, FGWHT BGRED "NO REPLY" RESETALL "\n");
			else
				fprintf(stdout, FGWHT BGGRN "%5d ms" RESETALL "\n", t / 1000);
		}
		// Pace ourselves.
		const long NS_PER_US = 1000L;
		const long int us_left = (long int) (timeout.tv_sec * 1000000L + timeout.tv_usec);
		struct timespec ts = {us_left / 1000000L, us_left * NS_PER_US};
		ts.tv_nsec = ts.tv_nsec % 1000000000L;
		nanosleep(&ts, 0);
	}
	fprintf(stdout, CLEARSCREEN);
	return 0;
}

// vi: tabstop=8


icmp_watch.c
// icmp_watch.c
//
// By Abraham Stolk.
// MIT LICENSE

#include <stdio.h>
#include <errno.h>
#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <termios.h>	// for tcsetattr()
#include <netdb.h>	// for getattrinfo()
#include <time.h>	// for nanosleep()
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip_icmp.h>
#include <netinet/icmp6.h>
#include <arpa/inet.h>
#include <sys/select.h>
#include <getopt.h>		// for getopt_long()

// NOTE:
// To allow root to use icmp sockets, run:
// $ sysctl -w net.ipv4.ping_group_range="0 0"
// To allow all users to use icmp sockets, run:
// $ sysctl -w net.ipv4.ping_group_range="0 2147483647"

#define ESC	    "\x1B"

#define RESETALL    ESC "[0m"

#define CURSORHOME  ESC "[H"

#define CLEARSCREEN ESC "[H" ESC "[2J" ESC "[3J"

#define FGWHT	    ESC "[1;37m"

#define BGRED	    ESC "[1;41m"

#define BGGRN	    ESC "[1;42m"

struct destination_info {
	int response_time;		// Response time in milliseconds
	int error;				// errno if there was an error, 0 otherwise
	struct sockaddr *address;			// pointer to either an sockadder_in or sockaddr_in6 struct
};

static struct termios orig_termios;    // The terminal settings before we modified it.

void disableRawMode()
{
	tcsetattr(STDIN_FILENO, TCSAFLUSH, &orig_termios);
}

void enableRawMode()
{
	tcgetattr(STDIN_FILENO, &orig_termios);
	atexit(disableRawMode);
	struct termios raw = orig_termios;
	raw.c_lflag &= ~(ECHO);	     // Don't echo key presses.
	raw.c_lflag &= ~(ICANON);    // Read by char, not by line.
	raw.c_cc[VMIN] = 0;	     // No minimum nr of chars.
	raw.c_cc[VTIME] = 0;	     // No waiting time.
	tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw);
}

static int ping_all(int cnt, struct destination_info* destinations, struct timeval* timeout)
{
	static int sequence = 0;
	const int seq = sequence++;        // the sequence number we will use for this run.
	const int waittime = (int) (timeout->tv_sec * 1000000 + timeout->tv_usec);
	int sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP);
	int sock6 = socket(AF_INET6, SOCK_DGRAM, IPPROTO_ICMPV6);
	if (sock < 0)
	{
		perror("socket");
		exit(4);
	}
	if (sock6 < 0) {
		perror("v6 socket");
		exit(6);
	}

	for (int i = 0; i < cnt; ++i) {
		destinations[i].response_time = -1;	  	// what we return if we did not get a reply.
		destinations[i].error = 0;			  	// what we return if there wasn’t an error
	}
	
	// Prepare an ICMP(v4 and v6) request for each destination.
	struct icmphdr icmp_hdr;
	unsigned char txdata[256];
	const int payloadsz = 10;
	memcpy(txdata + sizeof icmp_hdr, "icmp_watch", payloadsz);    // icmp payload
	
	struct icmp6_hdr icmp6_hdr;
	unsigned char tx6data[256];
	memcpy(tx6data + sizeof icmp6_hdr, "icmp_watch", payloadsz);    // icmpv6 payload

	// A fork in the road, as the code if different for v4 and v6
	for (int i = 0; i < cnt; ++i) {
		if (destinations[i].address->sa_family == AF_INET) {
			struct sockaddr_in addr;
			memset(&addr, 0, sizeof addr);
			addr.sin_family = AF_INET;
			addr.sin_addr = ((struct sockaddr_in *) destinations[i].address)->sin_addr;

			memset(&icmp_hdr, 0, sizeof icmp_hdr);
			icmp_hdr.type = ICMP_ECHO;
			icmp_hdr.un.echo.id = 0xbeef;

			icmp_hdr.un.echo.sequence = seq;
			memcpy(txdata, &icmp_hdr, sizeof icmp_hdr);
			int rc = sendto(sock, txdata, sizeof icmp_hdr + payloadsz, 0, (struct sockaddr*) &addr, sizeof addr);
			if (rc <= 0)
			{
				destinations[i].error = errno;
			}
		} else if (destinations[i].address->sa_family == AF_INET6) {
			struct sockaddr_in6 addr;
			memset(&addr, 0, sizeof addr);
			addr.sin6_family = AF_INET6;
			addr.sin6_addr = ((struct sockaddr_in6 *) destinations[i].address)->sin6_addr;

			memset(&icmp6_hdr, 0, sizeof icmp6_hdr);
			icmp6_hdr.icmp6_type = ICMP6_ECHO_REQUEST;
			icmp6_hdr.icmp6_id = 0xbeef;

			icmp6_hdr.icmp6_seq = seq;
			memcpy(tx6data, &icmp6_hdr, sizeof icmp6_hdr);
			int rc = sendto(sock6, tx6data, sizeof icmp6_hdr + payloadsz, 0, (struct sockaddr*) &addr, sizeof addr);
			if (rc <= 0)
			{
				destinations[i].error = errno;
			}
		}
	}

	int highestfd;
	if (sock > sock6)
		highestfd = sock;
	else
		highestfd = sock6;

	int num_replies = 0;
	while (num_replies < cnt)
	{
		// Set the file descriptor set for select()
		// According to the man page for select(), this should be done every iteration of a loop
		fd_set read_set;
		FD_ZERO(&read_set);
		FD_SET(sock, &read_set);
		FD_SET(sock6, &read_set);
		
		// wait for a reply with a timeout
		const int rc0 = select(highestfd + 1, &read_set, NULL, NULL, timeout);
		if (rc0 == 0)
		{
			// Timed out without a reply.
			close(sock);
			close(sock6);
			sock = 0;
			sock6 = 0;
			return num_replies;
		}
		else if (rc0 < 0)
		{
			close(sock);
			close(sock6);
			sock = 0;
			sock6 = 0;
			perror("select");
			return -1;
		}
		
		if (FD_ISSET(sock, &read_set)) {
			/* IPv4 code */
			unsigned char rcdata[256];
			struct icmphdr rcv_hdr;
			struct sockaddr_in other_addr;
			socklen_t other_addr_len = sizeof(other_addr);
			
			const int rc1 = recvfrom(sock, rcdata, sizeof rcdata, 0, (struct sockaddr*) &other_addr, &other_addr_len);
			if (rc1 <= 0)
			{
				perror("recvfrom");
				exit(5);
			}
			if (rc1 < (int) sizeof(rcv_hdr))
				exit(6);			// ICMP packet was too short.
			memcpy(&rcv_hdr, rcdata, sizeof rcv_hdr);
			assert(rcv_hdr.type == ICMP_ECHOREPLY);
			if (rcv_hdr.un.echo.sequence == seq)	// The sequence number should match, otherwise it's not a valid response.
			{
				const struct in_addr send_addr = other_addr.sin_addr;
				int idx = -1;
				// Look up which host sent us this reply.
				for (int j = 0; j < cnt; ++j)
					if(((struct sockaddr_in* ) destinations[j].address)->sin_addr.s_addr == send_addr.s_addr) {
						idx = j;
						break;
					}
				assert(idx >= 0);
				const int timeleft = (int) (timeout->tv_sec * 1000000 + timeout->tv_usec);
				destinations[idx].response_time = waittime - timeleft;
				num_replies += 1;
			}
		}
		if (FD_ISSET(sock6, &read_set)) {
			/* IPv6 code */
			unsigned char rcdata6[256];
			struct icmp6_hdr rcv6_hdr;
			struct sockaddr_in6 other_addr6;
			socklen_t other_addr6_len = sizeof(other_addr6);

			const int rc1v6 = recvfrom(sock6, rcdata6, sizeof rcdata6, 0, (struct sockaddr*) &other_addr6, &other_addr6_len);
			if (rc1v6 <= 0) {
				perror("recvfrom (v6)");
				exit(5);
			}
			if (rc1v6 < (int) sizeof(rcv6_hdr))
				exit(8);
			
			memcpy(&rcv6_hdr, rcdata6, sizeof rcv6_hdr);
			assert(rcv6_hdr.icmp6_type == ICMP6_ECHO_REPLY);
			if (rcv6_hdr.icmp6_seq == seq) {
				// The sequence number should match, otherwise it's not a valid response
				const struct in6_addr send6_addr = other_addr6.sin6_addr;
				int idx = -1;
				// Search for which host sent the reply
				for (int j = 0; j < cnt; ++j) {
					// For IPv6 we have to compare all 16 unsigned chars of the address (128 bits)
					int thisone = 1;
					for(int x = 0; x < 16; x++) {
						if(((struct sockaddr_in6* ) destinations[j].address)->sin6_addr.s6_addr[x] != send6_addr.s6_addr[x]) {
							// One of the bytes don't match, so it isn't this host
							thisone = 0;
							break;
						}
					}
					if (thisone) {
						idx = j;
						break;
					}
				}
				assert(idx >= 0);
				const int timeleft = (int) (timeout->tv_sec * 1000000 + timeout->tv_usec);
				destinations[idx].response_time = waittime - timeleft;
				num_replies += 1;
			}
		}
	}
	
	if (close(sock) < 0)
		perror("close(socket)");
	sock = 0;
	
	if (close(sock6) < 0)
		perror("close(socket for v6)");
	sock6 = 0;
	return num_replies;
}

// Resolves a set of hostnames to IPv4 or IPv6 addresses
static int get_ip_addresses(int cnt, char** hosts, int args_left, struct destination_info* destinations, int restrict_protocol)
{
	struct addrinfo hints;
	memset(&hints, 0, sizeof(hints));
	
	// Restrict protocol: if 0, both IPv4 and IPv6 are enabled, if 4, only v4 and if 6, only v6
	switch (restrict_protocol) {
		case 0:
			hints.ai_family = AF_UNSPEC;
			break;
		case 4:
			hints.ai_family = AF_INET;
			break;
		case 6:
			hints.ai_family = AF_INET6;
			break;
		default:
			fprintf(stderr, "Unknown protocol restrict of %i, setting to AF_UNSPEC", restrict_protocol);
			hints.ai_family = AF_UNSPEC;
			break;
	}
	hints.ai_socktype = SOCK_DGRAM;
	for (int i = 0; i < cnt; ++i)
	{
		struct addrinfo* infos;
		const char* host = hosts[args_left + i];
		const int rv = getaddrinfo(host, 0, &hints, &infos);
		if (rv)
		{
			fprintf(stderr, "getaddrinfo() %s\n", gai_strerror(rv));
			return i;
		}
		for (struct addrinfo* inf = infos; inf != 0; inf = inf->ai_next)
		{
			int haveaddr = 0;
			// Check if the returned address was v4 or v6
 			switch(inf->ai_family) {
				case AF_INET: {
					// Copy the inf->ai_addr structure
					// It'll need to be freed later with something like free(destinations->address)
					struct sockaddr_in* copied_ai_addr = malloc(sizeof(struct sockaddr_in));
					if(copied_ai_addr == NULL) {
						// Check we actually got some memory, if not, exit
						fprintf(stderr, "Failed allocate memory for copied_ai_addr!\n");
						exit(EXIT_FAILURE);
					}
					memcpy(copied_ai_addr, inf->ai_addr, sizeof(struct sockaddr_in));
					destinations[i].address = (struct sockaddr*) copied_ai_addr;
					haveaddr = 1;
					break;
				}
				case AF_INET6: {
					// Copy the inf->ai_addr structure
					// It'll need to be freed later with something like free(destinations->address)
					struct sockaddr_in6* copied_ai_addr = malloc(sizeof(struct sockaddr_in6));
					if(copied_ai_addr == NULL) {
						// Check we actually got some memory, if not, exit
						fprintf(stderr, "Failed allocate memory for copied_ai_addr!\n");
						exit(EXIT_FAILURE);
					}
					memcpy(copied_ai_addr, inf->ai_addr, sizeof(struct sockaddr_in6));
					destinations[i].address = (struct sockaddr*) copied_ai_addr;
					haveaddr = 1;
					break;
				}
				default: {
					fprintf(stderr, "For %s, we got an address type (%i) that wasn't AF_INET (%i) or AF_INET6 (%i)", host, inf->ai_family, AF_INET, AF_INET6);
				}
			}
			if (haveaddr) {
				// Break out of the loop if we have an address
				break;
			}
		}
		freeaddrinfo(infos);
	}
	return cnt;
}

void print_help(char *progname) {
	// Restyled the help to more closely match the --help text for mv, etc.
	printf("Usage: %s [option]... destination_ip...\n"
		   "Send batch requests for ICMP and show the results\nPress q or escape to exit\n\n"
		   "  -i, --interval=INTERVAL\tspecify how long in seconds to wait for replies (real numbers, e.g. 1.5 are allowed)\n"
		   " -4\t\t\t\tOnly use IPv4\n"
		   " -6\t\t\t\tOnly use IPv6\n"
		   "  -h, --help\t\t\tshow this help\n", progname);
}

int main(int argc, char* argv[])
{
	struct timeval default_timeout = {1, 0};    // seconds, microseconds.
	
	// Restrict protocol: if 0, both IPv4 and IPv6 are enabled, if 4, only v4 and if 6, only v6
	int restrict_protocol = 0;
	
	// Parse command line options (we'll break out of the loop)
	while(1) {
		int option_index = 0;
		int c;
		static struct option long_options[] = {
			{"interval", required_argument, 0, 'i'},
			{"help", no_argument, 0, 'h'},
			{0, 0, 0, 0} 	// default option (for unknown options)
		};
		
		c = getopt_long(argc, argv, "46i:h", long_options, &option_index);
		
		if (c == -1) {
			break; // no more options
		}
		
		//printf("Right now c is %i (%c)\n", c, c);
		switch(c) {
			case 0:
				printf("Unknown option %s\n", long_options[option_index].name);
				break;
			case 'i':
				// Convert the argument to --interval to a (double) float, then convert it to a timespec struct
				if(optarg) {
					double interval_double;
					errno = 0;		// Set errno to 0 so we can detect errors
					interval_double = strtod(optarg, NULL);
					if(errno != 0) {
						perror("Converting interval");
						exit(1);
					}
					// Set the specified interval as the default timeout
					default_timeout.tv_sec = (int) interval_double;
					default_timeout.tv_usec = (int) (interval_double * 1000000);
					break;
				} else {
					fprintf(stderr, "-i/--interval needs an argument\n");
					exit(1);
				}
			case '4':
				// Only use IPv4
				if (restrict_protocol == 0) {
					restrict_protocol = 4;
				} else {
					fprintf(stderr, "Only one of -4 or -6 is accepted\n");
					exit(1);
				}
				break;
			case '6':
				// Only use IPv6
				if (restrict_protocol == 0) {
					restrict_protocol = 6;
				} else {
					fprintf(stderr, "Only one of -4 or -6 is accepted\n");
					exit(1);
				}
				break;
			case 'h':
				print_help(argv[0]);
				exit(0);
			default:
				//fprintf(stderr, "getopt_long returned character code 0x%x\n", c);
				;
		}
	} // Finished parsing command line options

	if (optind < argc) {
		//printf("%i non-option argv elements?\n", argc - optind);
	} else {
		print_help(argv[0]);
		return 1;
	}

	const int cnt = argc - optind;    // Every argument left after taking away the options is a hostname.
	struct destination_info destinations[cnt];

	fprintf(stderr, "Looking up %d ip numbers...", cnt);
	fflush(stderr);
	const int num = get_ip_addresses(cnt, argv, optind, destinations, restrict_protocol);
	fprintf(stderr, "DONE\n");
	if (num != cnt)
	{
		fprintf(stderr, "Could not resolve all hostnames. Aborting.\n");
		exit(2);
	}

	enableRawMode();    // Don't echo keyboard characters, don't buffer them.

	int done = 0;
	int spaceForHostname = 19;	// Default spacing is 19 characters. (it was 20 before, but it'll get incremented later to ensure that there's always at least one space between the hostname and the result)
	
	// Check if any of the hostnames are longer than 20 characters and up spaceForHostname if the are
	for (int i = 0; i < cnt; i++) {
		/* this line will need changing if the cli options patch is merged */
		int lengthOfThisHostname = strlen(argv[1 + i]);
		if(lengthOfThisHostname > spaceForHostname) {
			spaceForHostname = lengthOfThisHostname;
		}
	}
	// Increment space for hostname by one so that there will always be at least one space between it and the result
	spaceForHostname++;
	
	while (!done)
	{
		// When ESC or Q is pressed, we should terminate.
		char c;
		const int numr = read(STDIN_FILENO, &c, 1);
		if (numr == 1 && (c == 27 || c == 'q' || c == 'Q'))
			done = 1;
		struct timeval timeout = default_timeout;
		ping_all(cnt, destinations, &timeout);
		fprintf(stdout, CLEARSCREEN);
		for (int i = 0; i < cnt; ++i)
		{
			const int t = destinations[i].response_time;
			const int e = destinations[i].error;
			fprintf(stdout, "%-*s", spaceForHostname, argv[optind + i]);
			if (t < 0)
				if (e != 0)
					fprintf(stdout, FGWHT BGRED "   ERROR" RESETALL " (%s)\n", strerror(e));
				else
					fprintf(stdout, FGWHT BGRED "NO REPLY" RESETALL "\n");
			else
				fprintf(stdout, FGWHT BGGRN "%5d ms" RESETALL "\n", t / 1000);
		}
		// Pace ourselves.
		const int NS_PER_US = 1000;
		const int us_left = (int) (timeout.tv_sec * 1000000 + timeout.tv_usec);
		struct timespec ts = {0, us_left * NS_PER_US};
		nanosleep(&ts, 0);
	}
	fprintf(stdout, CLEARSCREEN);
	
	// Free addresses got from get_ip_addresses
	for (int i = 0; i < cnt; i++) {
		free(destinations[i].address);
	}
	
	return 0;
}

// vi: tabstop=8
